### PR TITLE
[CBRD-22793] Remove the legacy REGEXP library

### PR DESCRIPTION
**Pull request by hgryoo**
http://jira.cubrid.org/browse/CBRD-22793

ECMAScript grammar is used to construct REGEX object. The ECMAScript in <regex> covers the POSIX ERE which supported in the old library. In addition, it supports more features than the POSIX.

This is an in-progress PR. Following are task lists:
- replace the old library with the C++11 <regex>
  * [x] remove including the old library's header
  * [x] remove all variables from the old library
  * [x] remove the unused the regex_matches function
  * [x] re-implement functions to use <regex>

TODO in other PR's
- unicode aware enhancement

### CODE REVIEW

**Comment by hgryoo** [comment]
@eseokoh, could you please give me any advice on the following?

**A bug of regex icase option with devtoolset-6**

With devtoolset-6, \<regex\> has inconsistent behavior with icase option. The icase option makes the whole pattern case-insensitive. For example, `aBc` would match `AbC` with icase option.
Following are some tests with gcc 6.
- `SELECT 'top' rlike '^[A-Z]{2,6}';` : false -- it should be true because rlike is case-insensitive.
- `SELECT 'TOP' rlike '^[A-Z]{2,6}';` : true -- absolutely
- `SELECT 'aaa' rlike '^[A-Z]{2,6}';` : true -- case-insensitive works here.
- `SELECT 'AAA' rlike '^[A-Z]{2,6}';` : true
- `SELECT 'bbb' rlike '^[A-Z]{2,6}';` : false -- ?

It seems that the icase option only works on first character in a range.
And I found that all of these examples works well with gcc 7.


**Modifying test cases**

We need to modify some test cases because of the library replacement.
1) word boundary syntax changed from `[[:<:]]` and `[[:>:]]` to `\b` 
2) lazy quantifier(`??`) is supported in \<regex\>

````
SELECT ( cast( _utf8'Öö_Şş_Üü2' as string charset iso88591)  REGEXP  cast( _utf8'Öö_Şş' as string charset iso88591));  
````

In the above case, Şş becomes `??` and it caused the repetition-operator invalid error before.
Now, Lazy quantifier(`??`) is supported in <regex> and we don't need to check it. 
Any other test cases for unicode will be covered in other PR.

**Allocator for std::regex**

Regex patterns might be re-used when REGEXP is in where clause. so pre-compilation of regex pattern reduces execution time of REGEXP.
Because std::regex is not allocator-aware, I have implemented the pre-compilation with global allocator (new and delete) until now. Using an instance of cubmem:private_allocator could be an option, but it might make code messy.

**Comment by eseokoh** [comment]
please refer to http://jira.cubrid.org/browse/CBRD-22854 for gcc8 upgrade.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/query/string_opfunc.c (1 incorrect indentation(s) found)
```
  4458     std::string src (src_char_string_p, src_length);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by hgryoo** [comment]
I would like to start a review but I'm going to continue tasks described in the above.
During the review, I would like to discuss the following.
**std::regex is not allocator-aware**
````
using my_string = basic_string <char, std::char_traits<char>, private_allocator<char>> // ok
using my_regex = basic_regex <char, std::regex_traits<char>, private_allocator<char>> // error, there is no template argument for allocator.
````
It means that the custom allocator has to be instantiated whenever constructing or destructing an std::regex object.
````
std::regex *compiled_reg;

//allocation and construction
{
cubmem::private_allocator<std::regex> alloc;
std::regex r(pattern, reg_flags);
compiled_reg= alloc.allocate(1);
alloc.construct(compiled_reg, r);
}
~~~
//desctruction and deallocation
{
cubmem::private_allocator<std::regex> alloc;
alloc.destroy(compiled_reg);
alloc.deallocation(compiled_reg, 1);
}
````

To avoid this the messy use of allocator and to make easy to manage the pointer, in my opinion, we need to support the following.

**in memroy_private_allocator.hpp**
- private_allocator doesn't have a construct function with arguments
````
template< class U, class... Args >
void construct( U* p, Args&&... args );
````
- private_unique_ptr looks like wrapper of unique_ptr. private_unique_ptr needs the following to use easily
	* default constructor to assign later.
	* bool operator to check whether the smart_ptr has nullptr

**Comment by eseokoh** [comment]
I don't think regex needs private allocation.

**Comment by hgryoo** [comment]
Thank you for your comment. db_private_alloc was used to construct the henry's regex. Do you mean that default allocation(new and delete) is enough?

**Comment by hgryoo** [comment]
the cci driver also uses the old regex library to match URL (cci_url_match in cci_util.c). But I'm not sure whether it has to be resolved in this issue.

**Comment by eseokoh** [comment]
@hgryoo 
why not?

**Comment by hgryoo** [comment]
@eseokoh, I removed the external folder and CMake script for the libregex. the cci_url_match also re-implemented with <regex>. I'm ready to be reviewed.

**Comment by eseokoh** [comment]
Have anyone of your team reviewed this?

**Comment by hgryoo** [comment]
No. I thought that one of my team members would be included as a reviewer in this PR. I'll let you know after doing team review.

**Comment by hgryoo** [comment]
@eseokoh, team review has been finished

**Comment by hgryoo** [comment]
Because of wrong pushed commits, I reset to the last reviewed commit. sorry for the mistake.

File: src/cci/cci_util.c [hunk 1]
```diff
@@ -994,79 +994,68 @@ cci_url_match (const char *src, char *token[])
...
+  }
+  // *INDENT-ON*
 
   if (error != CCI_ER_NO_ERROR)
     {
       /* free allocated memory when error was CCI_ER_NO_MORE_MEMORY */
-      for (i = 0; match_idx[i] != -1 && match[match_idx[i]].rm_so != -1; i++)
+      for (int i = 0; match_idx[i] != -1; i++)
 	{
 	  FREE_MEM (token[i]);
```
Comment (shparkcubrid) [inline src/cci/cci_util.c]: it can FREE_MEM() for token which is uninitialized.

Comment (shparkcubrid) [inline src/cci/cci_util.c]: it can FREE_MEM() for token which is uninitialized.

File: src/query/string_opfunc.c [hunk 2]
```diff
@@ -4580,33 +4490,30 @@ db_string_rlike (const DB_VALUE * src_string, const DB_VALUE * pattern, const DB
...
+    }
+  catch (std::regex_error & e)
+    {
+      // regex execution exception, error_complexity or error_stack
       error_status = ER_REGEX_EXEC_ERROR;
-      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error_status, 1, rx_err_buf);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error_status, 1, e.what ());
       *result = V_ERROR;
-      break;
+      goto cleanup;
```
Comment (byungwook-kim) [inline src/query/string_opfunc.c]: "goto cleanup" might not be needed because the code would be reached to "cleanup".

**Pull request merged**